working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+  // access to the original `s` and mutations performed in this
+  // isolate will be visible to other isolates.
+  // 
+  // (Subject to memory model constraints).
+  s.v = 10; 
+});
+
+expect(equals(10), s.v);
+```
+
+To make state sharing between isolates simpler we also allow to declare static fields which are shared between all isolates in the isolate group. `@shared` global fields are required to have static type that is a subtype of `Shareable?`. 
+
+```dart 
+// All isolates within an isolate group share this variable.
+@shared int v = 0;
+```
+
+### Functions
+
+Given the parallel between functions and other objects it makes sense to declare that functions which close only over shareable state are themselves shareable.
+
+**Static type of a function literal which captures only `Shareable` variables is  `ShareableFunction<F extends Function>` which is subtype of both `F` and `Shareable`.**
+
+```dart
+void main() {
+  final Shareable s = ...;
+  final f = () => s;  // ShareableFunction<Shareable Function()>
+
+  final Object o;
+  final g = () => o;  // Not shareable
+}
+
+class S implements Shareable {
+  void Function() f; // compile time error
+  
+  ShareableFunction<void Function()> f;  // ok
+}
+```
+
+Shareable functions have one subtle behavior - they capture and share local variables, but not the global state of the isolate. Consider for example the following code:
+
+``` dart
+int global = 0; 
+
+void main() async {
+  final s = S();
+  
+  s.v = 42;
+  global = 42;
+  
+  await Isolate.run(() {
+    print(global);  // => 0
+    global = 24;
+
+    print(s.v);  // 42
+    s.v = 24;
+  });
+  
+  print(global);  // 42
+  print(s.v);  // 24
+}
+```
+
+This behavior might be somewhat surprising to developers - especially if a functions depends on or updates some mutable global state. Whether this is a problem in practice is unclear: sending a closures through a `SendPort` leads to similar behavior.
+
+### Records
+
+Records where all components are shareable are themselves shareable. 
+
+### Collections (`Iterable`, `List`, `Set`, `Map`)
+
+If `S` shareable then `Iterable<S>`, `List<S>` and `Set<S>` are shareable. If `SK` and `SV` are sharable then `Map<SK, SV>` is shareable. This imposes additional requirements on classes implementing these interfaces: whether some container is shareable depends on concrete type parameters of an instance. If class `C<T1, ..., Tn>` implements `List<F(T1, ..., Tn)>` then constructing an instance of `C<X1, ..., Xn>` , such that `F(X1, ..., Xn)` is shareable, should only be possible if `C<X1, ..., Xn>` satisfies restrictions imposed on shareable classes i.e. static type of each of its fields should be a subtype of `Shareable`. 
+
+``` dart
+class C1<T> implements List<T> {
+  final List<T> v;
+}
+
+class C2<T> implements List<T> {
+  final List<Object> v;
+}
+
+C1<Shareable>(); // ok
+C2<Shareable>(); // runtime error: static type of C2.v is not shareable
```

Comment:
Why not a compile time error?

Reply:
You are right that in this particular case it could have also been a compile time error, but in general case detecting this in compile time is not possible (or rather it is possible but requires significant changes to the typesystem to express this constraint). 

In any case I have remove this section entirely and went for a simpler variant.